### ISSUE TITLE
HTTP Connection Timeout

### ISSUE THREAD

**Issue by tustvold** [timberio/vector#5858]
### Current Vector Version

```
0.11.1
```

### Use-cases

Whilst investigating #5846 I noticed that the HTTPConnector is not set to have a connection timeout - see [here](https://github.com/timberio/vector/blob/master/src/http.rs#L52) and [here](https://docs.rs/hyper/0.13.9/hyper/client/connect/struct.HttpConnector.html#method.set_connect_timeout). Whilst it turns out not to be the issue causing #5846 in my case, it could theoretically cause similar symptoms.

### Attempted Solutions

I'm not sure there is a trivial way to set this currently

### Proposal

Add a global connection timeout option, potentially with a large default of 5 minutes, to set this on the HTTP client

### References

- #5846

**Comment by ChinYing-Li** [comment]
Hi, is this still a valid issue? If so, I would like to give it a try.

**Comment by jszwedko** [comment]
Hi @ChinYing-Li !

Yes, this is still something that we want. I think it should be configurable at the source level. We should probably also have a default timeout of, say, 5s.

**Comment by ChinYing-Li** [comment]
Thank you for the pointer. I suppose we should make the timout configurable for every sink and source?

**Comment by jszwedko** [comment]
Yeah, I think so. You can see some precedent for other configuration like `tls` and `request`:

https://github.com/vectordotdev/vector/blob/587baaac58ff97f078bea4f58d62a257fb5fe741/src/sinks/clickhouse.rs#L43-L44

If you want, I'd recommend starting with a single component first, just to get feedback, and then we can roll the pattern out to the rest.
